File: 19/day19.c
#include <stdio.h>

static int run(int acc) {
    static int const lim[] = { 974, 10551374 };

    int const n = lim[acc % 2];
    int /*j,*/ i;

    /* Translated machine code from my input
     * essentially boils down to this: */
    /*
    for (acc = 0, i = 1; i <= n; ++i)
        for (j = 1; j <= n; ++j)
            if (i * j == n)
                acc += i;
    */

    /* Slightly optimized: */
    for (acc = 0, i = 1; i <= n; ++i)
        if (n % i == 0)
            acc += n / i;

    return acc;
}

int main(void) {
    printf("%d\n%d\n", run(0), run(1));
    return 0;
}

File: day19.py
#! /usr/bin/env python3
from elfcode_emulator import *

def computed_value(register_values, iterations):
    m = Machine(program_text, register_values)

    try:
        for i in range(iterations):
            regs = next(m)
    except StopIteration:
        return regs[0]

    n = regs[3]

    # Translated machine code (from my input at least)
    # boils down to:
    #acc = 0
    #for i in range(n + 1):
    #    for j in range(1, n + 1):
    #        if i * j == n:
    #            acc += i

    # Here is a slight improvement though:
    return sum(0 if n % i else n // i for i in range(1, n + 1))

program_text = sys.stdin.read()

# This was discovered empirically by emulating the input Elfcode.
print(computed_value([ 0, 0, 0, 0, 0, 0 ],     3 * 3))
print(computed_value([ 1, 0, 0, 0, 0, 0 ], 1 + 4 * 4))
